### ISSUE TITLE
Kind error in Ganeti.Lens.traverseOf2

### ISSUE THREAD

**Issue by thielema** [ganeti/ganeti#1748]
Building on Ubuntu 24.04 Noble Numbat or NixOS 24.05 yields this strange kind error:

~~~~
src/Ganeti/Lens.hs:98:32: error: [GHC-25897]
    * Couldn't match kind `k2' with `*'
      When matching types
        Compose f g0 :: * -> *
        Compose f g :: k2 -> *
      Expected: s -> Compose f g t
        Actual: s -> Compose f g0 t0
      `k2' is a rigid type variable bound by
        the type signature for:
          traverseOf2 :: forall {k1} {k2} (f :: k1 -> *) (g :: k2 -> k1) s
                                (t :: k2) a (b :: k2).
                         LensLike (Compose f g) s t a b -> (a -> f (g b)) -> s -> f (g t)
        at src/Ganeti/Lens.hs:(96,1)-(97,45)
    * In the second argument of `(.)', namely
        `traverseOf k (Compose . f)'
      In the expression: getCompose . traverseOf k (Compose . f)
      In an equation for `traverseOf2':
          traverseOf2 k f = getCompose . traverseOf k (Compose . f)
    * Relevant bindings include
        f :: a -> f (g b) (bound at src/Ganeti/Lens.hs:98:15)
        k :: LensLike (Compose f g) s t a b
          (bound at src/Ganeti/Lens.hs:98:13)
        traverseOf2 :: LensLike (Compose f g) s t a b
                       -> (a -> f (g b)) -> s -> f (g t)
          (bound at src/Ganeti/Lens.hs:98:1)
   |
98 | traverseOf2 k f = getCompose . traverseOf k (Compose . f)
   |                                ^^^^^^^^^^^^^^^^^^^^^^^^^^
~~~~

Cannot reproduce this when building with `Cabal`, it only happens with `make`.

**Comment by thielema** [comment]
We can replace `Over (->)` by `LensLike`, which is simpler and closer to the type signature of `traverseOf`. However, this does not solve the kind error.

https://hackage.haskell.org/package/lens-5.3.2/docs/Control-Lens-Traversal.html#v:traverseOf

**Comment by saschalucas** [comment]
As a usual workaround removing the type declaration as in commit 8de9b74c5422fd1706bfade2d7ff13703aeeb1fd / PR #1716 will skip over this problem and a second one:

```
[GHC]: src/Ganeti/Lens.o <- cabal_macros.h src/Ganeti/Lens.hs

src/Ganeti/Lens.hs:104:29: error:
    • Couldn't match kind ‘k1’ with ‘*’
      When matching types
        g0 :: * -> *
        g :: k1 -> *
      Expected: s -> WrappedMonad m (g t)
        Actual: s -> WrappedMonad m (g0 a0)
      ‘k1’ is a rigid type variable bound by
        the type signature for:
          mapMOf2 :: forall {k1} (m :: * -> *) (g :: k1 -> *) s (t :: k1) a
                            (b :: k1).
                     Over (->) (Compose (WrappedMonad m) g) s t a b
                     -> (a -> m (g b)) -> s -> m (g t)
        at src/Ganeti/Lens.hs:(102,1)-(103,41)
    • In the second argument of ‘(.)’, namely
        ‘traverseOf2 k (WrapMonad . f)’
      In the expression: unwrapMonad . traverseOf2 k (WrapMonad . f)
      In an equation for ‘mapMOf2’:
          mapMOf2 k f = unwrapMonad . traverseOf2 k (WrapMonad . f)
    • Relevant bindings include
        f :: a -> m (g b) (bound at src/Ganeti/Lens.hs:104:11)
        k :: Over (->) (Compose (WrappedMonad m) g) s t a b
          (bound at src/Ganeti/Lens.hs:104:9)
        mapMOf2 :: Over (->) (Compose (WrappedMonad m) g) s t a b
                   -> (a -> m (g b)) -> s -> m (g t)
          (bound at src/Ganeti/Lens.hs:104:1)
    |
104 | mapMOf2 k f = unwrapMonad . traverseOf2 k (WrapMonad . f)
    |                             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```
However I'm unable to contribute to the root cause.

**Comment by thielema** [comment]
Here is an explanation:
https://mail.haskell.org/pipermail/haskell-cafe/2024-July/136843.html

Summary:
GHC-9.2 started to enable the language extension set `GHC2021` by default. Part of it is the extension `PolyKinds`. We can disable it by `{-# LANGUAGE NoPolyKinds #-}`. Cabal disables `GHC2021`, which is pretty sensible in order to make Cabal packages compile across different versions of GHC.

**Comment by saschalucas** [comment]
solved with #1759
